### DISCUSSION TITLE
An API using marp-cli is working on deployed to Docker, but not working on deployed to Heroku.

### DISCUSSION THREAD

**Discussion by link2004** [marp-team/marp#525] [Q&A] [answered]
> [!NOTE]
> The title of this thread has been translated into English by the forum moderator.

<details>
<summary>Original title</summary>

### ローカルで、marp-cliを使ったAPIをDockerでデプロイすると正常に動くのに、Herokuにデプロイするとうまく動かない。

</details>

---

I'm creating an API that takes Markdown as input and returns HTML, PDF, and PPTX.

Local environment OS: Linux (Ubuntu WSL)

### Local Deployment Procedure

Execute the following commands in the root directory:
`docker build -t marp-api .`
`docker run -p 3005:3005 marp-api`

Then, execute the curl command:
![image](https://github.com/user-attachments/assets/62eb1239-1b01-47d6-8252-9366c70be727)
All are functioning normally.

However, when this Docker container is deployed to Heroku, the API stops working correctly.

### Heroku Deployment Procedure

   ```bash
   # Heroku CLIにログイン
   heroku login

   # Herokuアプリを作成
   heroku create my-marp-app

   # HerokuのContainer Registryにログイン
   heroku container:login

   # Dockerイメージをビルド
   docker build -t my-marp-app .

   # DockerイメージをHerokuにプッシュ
   heroku container:push web -a my-marp-app

   # アプリをリリース
   heroku container:release web -a my-marp-app

   # アプリケーションのログを確認
   heroku logs --tail -a my-marp-app
   ```
The deployment was completed successfully.
However, when trying to create PDF or PPTX on the deployed API, an error occurs.

### Error content: In Generated PDF
`{"error":"Error generating PDF: Failed converting Markdown. (Failed to launch the browser process!\n[0726/092736.084575:FATAL:setuid_sandbox_host.cc(158)] The SUID sandbox helper binary was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. You need to make sure that /opt/google/chrome/chrome-sandbox is owned by root and has mode 4755.\n[0726/092736.084835:ERROR:scoped_ptrace_attach.cc(27)] ptrace: Permission denied (13)\n\n\nTROUBLESHOOTING: https://pptr.dev/troubleshooting\n)"}
`

### Error log from heroku logs --tail -a marp-api
`2024-07-26T09:27:36.090885+00:00 heroku[router]: at=info method=POST path="/pdf" host=marp-api-～～～～～.herokuapp.com request_id=3ccc801f-c698-4887-b18d-d0cc0876b2d1 fwd="114.163.143.8" dyno=web.1 connect=0ms service=292ms status=500 bytes=1542 protocol=https`

HTML can be downloaded normally.

### I think the problem is likely related to Chrome, but I don't know how to resolve it.
Is it possible for results to vary depending on the environment even when using Docker?

Directory structure:
![image](https://github.com/user-attachments/assets/a395d935-43d7-4427-a054-d2cc6a97c07e)

Main code api.js:
```
const express = require('express');
const fs = require('fs').promises;
const path = require('path');
const os = require('os');
const { marpCli } = require('@marp-team/marp-cli');
const cors = require('cors');
const helmet = require('helmet');
const rateLimit = require('express-rate-limit');

const app = express();
const port = process.env.PORT || 3005;

// Set security headers
app.use(helmet());

// Set rate limit
const limiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  max: 100 // 100 requests per IP address
});
app.use(limiter);

// Set CORS (allow specific origin in production)
app.use(cors({
  origin: process.env.ALLOWED_ORIGIN || 'http://localhost:3000'
}));

app.use(express.json({ limit: '1mb' })); // Limit JSON body size

// Custom CSS file path
const customCssPath = path.join(__dirname, "custom.css");

const generateFile = async (markdown, format, res) => {
  try {
    // Create temporary file
    const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), 'marp-'));
    const inputFile = path.join(tempDir, 'input.md');
    const outputFile = path.join(tempDir, `output.${format}`);

    await fs.writeFile(inputFile, markdown);

    // Generate file using marp-cli
    const cliOptions = [
      inputFile,
      '--output',
      outputFile,
      '--theme',
      customCssPath,
      `--${format}`
    ];
    await marpCli(cliOptions);

    const fileContent = await fs.readFile(outputFile);

    // Cleanup
    await Promise.all([
      fs.unlink(inputFile),
      fs.unlink(outputFile)
    ]);
    await fs.rmdir(tempDir);

    res.setHeader('Content-Type', format === 'html' ? 'text/html' : format === 'pdf' ? 'application/pdf' : 'application/vnd.openxmlformats-officedocument.presentationml.presentation');
    res.setHeader('Content-Disposition', `attachment; filename=marp-slide.${format}`);
    res.send(fileContent);
  } catch (error) {
    console.error(`Error generating ${format.toUpperCase()}:`, error);
    res.status(500).json({ error: `Error generating ${format.toUpperCase()}: ${error.message}` });
  }
};

app.post('/:format(html|pdf|pptx)', async (req, res) => {
  const { markdown } = req.body;
  const { format } = req.params;

  if (!markdown) {
    return res.status(400).json({ error: 'Markdown content is required' });
  }

  await generateFile(markdown, format, res);
});

app.listen(port, () => {
  console.log(`Server running on port ${port}`);
});
```

Dockerfile
```
# Specify base image
FROM node:16

# Create working directory
WORKDIR /usr/src/app

# Install necessary tools
RUN apt-get update && apt-get install -y \
    wget \
    gnupg \
    --no-install-recommends && \
    rm -rf /var/lib/apt/lists/*

# Manually download and install deb file
RUN wget -q https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb && \
apt-get update && \
apt-get install -y ./google-chrome-stable_current_amd64.deb && \
rm ./google-chrome-stable_current_amd64.deb

# Copy package.json and package-lock.json
COPY package*.json ./

# Install dependencies
RUN npm install

# Copy application source code
COPY . .

# Specify application port
EXPOSE 3005

# Start application
CMD ["node", "api.js"]

```

package.json
```
{
  "name": "marp-api",
  "version": "1.0.0",
  "main": "api.js",
  "scripts": {
    "build": "npm install",
    "start": "node api.js"
  },
  "dependencies": {
    "@marp-team/marp-cli": "^3.4.0",
    "cors": "^2.8.5",
    "express": "^4.19.2",
    "express-rate-limit": "^7.3.1",
    "helmet": "^7.1.0",
    "puppeteer-core": "^22.14.0"
  }
}

```

heroku.yml
```
build:
  docker:
    web: Dockerfile

```

[The repository for this API is in this public branch](https://github.com/link2004/marp-ai-api/tree/public)

**Comment by yhatt** [comment]
> [!NOTE]
> @link2004 For the Marp community's discoverability and workability, please your help in writing the description in English. You can also use a translator.

**Reply by link2004** [reply]
sorry! I've just updated it to English, so please check it out.

**Reply by yhatt** [reply]
The title of this discussion also has been translated into English by the forum moderator: `An API using marp-cli is correctly working on deployed to Docker, but not working on deployed to Heroku.`

**Comment by yhatt** [comment, marked as answer]
```
The SUID sandbox helper binary was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. You need to make sure that /opt/google/chrome/chrome-sandbox is owned by root and has mode 4755.
```

This error is about Chrome's [process sandboxing](https://chromium.googlesource.com/chromium/src/+/main/docs/design/sandbox.md).

Marp CLI is [automatically detecting whether it was executed inside a container](https://github.com/marp-team/marp-cli/blob/37cd0084ba9bed7a7fa1714e48a59ee4986c56cc/src/utils/puppeteer.ts#L79) such as Docker and Podman (powered by [`is-inside-container`](https://github.com/sindresorhus/is-inside-container) package), and disable incompatible process sandboxing automatically. But this detection may not work in some PaaS platform like Heroku.

You can try to add `CHROME_NO_SANDBOX` environment value into your container image as `1`, to disable Chrome's process sandboxing forcibly.

```dockerfile
ENV CHROME_NO_SANDBOX=1
```

**Reply by link2004** [reply]
It worked!!! Thank you so much. My three-day struggle was resolved in one minute.
